src/units/animation.cpp

```diff
@@ -1329,6 +1373,196 @@ void unit_animator::add_animation(unit_const_ptr animated_unit
+		} else {
+			m_u.set_location(map_location::null_location());
+		}
+		disp->invalidate(u_initial_loc);
+		m_dst = find_vacant_tile(m_dst, pathfind::VACANT_ANY, m_up.get());
+		config cfg;
+		cfg["type"] = m_u.type().id();
+		cfg["gender"] = m_u.gender();
+		cfg["variation"] = m_u.variation();
+		cfg["side"] = m_u.side();
+		cfg["x"] = std::to_string(m_src.wml_x()) + "," + std::to_string(m_dst.wml_x());
+		cfg["y"] = std::to_string(m_src.wml_y()) + "," + std::to_string(m_dst.wml_y());
+		cfg["force_scroll"] = true;
+		vconfig vcfg = vconfig(cfg);
+		m[tag]({"","",map_location(),map_location(),config()},vcfg);
+		result = screen_p->maybe_rebuild();
+		if (!result){
+			screen_p->invalidate_all();
+		}
+		m_u.set_location(u_initial_loc);
+		index_movement_anim++;
+	}
+	disp->update();
+	return index_movement_anim < animated_units_.size() && animated_units_[index_movement_anim].is_movement;
+}
+
+bool unit_animator::move_units_fake(int& index, const int& size)
+{
+	if(size < 1) {
+		if(index < animated_units_.size()) {
+			return animated_units_[index].is_movement;
+		} else {
+			return false;
+		}
+	}
+	const int start = index;
+	int temp_index = index;
+	const std::string tag = "move_units_fake";
+	game_events::wml_action::map m = game_events::wml_action::registry(); // m[tag] is call wml func by tag.
+	game_display* screen_p = game_display::get_singleton();
+	display* disp = display::get_singleton();
+	bool result;
+	bool coherent_moving = false;
+	std::vector<std::vector<std::reference_wrapper<anim_elem>>> processed_anim_queues;
+
+	while(temp_index < animated_units_.size() && temp_index < start + size) {
+		auto& anim = animated_units_[temp_index];
+		// either create a new vector when fail to match existing unit, or append to last matched unit vector.
+		bool new_queue = true;
+		for(auto iter = processed_anim_queues.begin(); iter != processed_anim_queues.cend(); iter++) {
+			auto& sorted_anim = (*iter)[0].get();
+			if(sorted_anim.move_up == anim.move_up) {
+				(*iter).emplace_back(std::ref(anim));
+				new_queue = false;
+				break;
+			}
+		}
+		if(new_queue) {
+			processed_anim_queues.emplace_back(std::vector<std::reference_wrapper<anim_elem>>{std::ref(anim)});
+		}
+		temp_index++;
+	}
+	anim_elem* last_anim_p = nullptr;
+	std::vector<std::vector<config>> uconfig_queues;
+	std::vector<std::pair<std::reference_wrapper<unit_ptr>, map_location>> initial_uloc_pairs;
+	// generate configs judged if the moving can be coherent
+	for(auto& vec : processed_anim_queues) {
+		std::vector<config> judged_uconfigs;
+		for(auto iter = vec.begin(); iter != vec.cend(); iter++) {
+			auto& anim = (*iter).get();
+			coherent_moving
+				= last_anim_p && anim.coherence && last_anim_p->move_up == anim.move_up && last_anim_p->src == anim.src;
+			unit_ptr& m_up = anim.move_up;
+			unit& m_u = *m_up;
+			map_location& m_src = coherent_moving ? last_anim_p->move_dst : anim.src;
+			map_location& m_dst = anim.move_dst;
+
+			bool pair_found = false;
```

Comment:
See my comment below. Constructing a separate `initial_uloc_pairs` list might not be necessary. Instead, can you handle `set_standing` here?